### ISSUE TITLE
Add ALT-Tag to banner and profile pictures and make the value editable for acessiblility reasons 

### ISSUE THREAD

**Issue by tigexplorer** [tootsuite/mastodon#10313]
### Pitch

The banner and profile picts on the profile site has no ALT or ARIA tags. Blind people can't get information about the content of this images. 
### Motivation

Additional text information for images are a basic necessity of the accessibility for handicap users.

**Comment by rkingett** [comment]
I would very much love this to be standard. I'd say make it not ARIA dependent because I'm unsure if ARIA works for sighted users that have images disabled.

**Comment by mgifford** [comment]
It isn't uncommon to have the profile picture just to be someone's name. Not as good as a description of the photo/image, but generally that is ok. 

The hero image definitely should have an option for alt text.

**Comment by rkingett** [comment]
I think Glitch Social or Hometown was working on this, but I agree.

**Comment by mgifford** [comment]
I don't know if Mastodon has a way to easily hide content from sighted users, but display it to keyboard only users but this is what Drupal has done:

https://www.drupal.org/docs/accessibility/hide-content-properly

This allows you to create a div with a background image and then put hidden text behind it.
```
<div class="background-image">
  <span class="visually-hidden">Textual alternatives for the background image</span>
</div>
```

**Comment by renchap** [comment]
If someone wants to work on this, please tell us and we will help you to get it done.

This work would involve:
- adding columns for the alt text in the `Account` model
- update the ActivityPub serializers to add the value to the AP representation, similar to what we do for status media
- ensure that it is saved in the database when received from ActivityPub
- update the web UI to add the `alt` tag to banners and avatars. We may want to not display it, or use an `aria` attribute, if the avatar is used to trigger an account hover card
- update the profile screen to allow the user to change those values

**Comment by ThisIsMissEm** [comment]
As per #32538 we'd want to use `summary` instead of `name` for the description at ActivityPub level

**Comment by mayakarabula** [comment]
Hey, I will be happy to work on this, I can take a look today in the evening.

With the caveat, that I have worked with ruby on rails, react, postgres but not with ActivityPub but I would be happy to learn if that's ok.

**Comment by ThisIsMissEm** [comment]
@mayakarabula that's perfectly fine! If you want I can jump on a call & walk you through the code & help you get setup.

I'd probably suggest starting with adding the columns + the settings page textareas & controller changes (I think there's validation on alt text length), then moving on to API and ActivityPub (incoming) and outgoing.

**Comment by ThisIsMissEm** [comment]
> update the web UI to add the alt tag to banners and avatars. We may want to not display it, or use an aria attribute, if the avatar is used to trigger an account hover card

@renchap I think in feeds and status pages, we'd probably not want to have the alt text on the profile picture, but just on the profile pages, where you'd typically get it. Otherwise we might be providing too much irrelevant information to people using screenreaders.

That is, focus on what is the most important text & media on the page.

@stefanbohacek does this make sense from an accessibility point of view?

**Comment by stefanbohacek** [comment]
@ThisIsMissEm I think this sounds good, but I am not really an expert, so I hope @ericwbailey and @rkingett won't mind if I tag them.

**Comment by rkingett** [comment]
quoting,

> I think in feeds and status pages, we'd probably not want to have the alt text on the profile picture, but just on the profile pages, where you'd typically get it.

Exactly this! It takes a lot of time to read the TL as is with a screen reader and I don't want profile picture alt text in any timeline, only profile pages!

**Comment by mayakarabula** [comment]
@ThisIsMissEm 
Hey, thanks for offering help with a walkthrough! I have sent you a message on mastodon

**Comment by ericwbailey** [comment]
Thanks for tagging me, @stefanbohacek. I really like this idea, as well as the considerations for where and how this information gets exposed.

That said, I am wary of contributing to this project until there are better guarantees about how accessibility feature requests are needed, especially when they address normative WCAG violations. I have observed how these sorts of things are handled, especially when they conflict with aesthetic notions.

**Comment by AlainVagner** [comment]
Maybe we should analyse a little bit what would be interesting to be done here before starting implementing stuff.

A profile pic is mainly decorative, and the information about who is writing is already available as a text nearby. But most of the occurrences of these avatar images are inside links or buttons (some links should probably be replaced by buttons, for example on the profile page). If the image is in a button we should check what is the accessible name of this button, which should state what is the action done by the button.
Here is an example of the current implementation in Mastodon 4.3:
```
<a class="avatar" href="https://[...].jpg" rel="noopener noreferrer" target="_blank">
    <div class="account__avatar" style="width: 90px; height: 90px;">
        <img src="[...].jpg" alt="">
    </div>
</a>
```
There is an event handler on the link, opening the image into a modal. This link should for me be a button, it does not lead to another webpage but does an action on the page. This link is empty, there is no label in this link. A user of a screen reader will not know where this link leads and will not know that it will show the image in a modal window.

We could have something instead like this:
```
<button class="avatar"><img src="..." alt="show profile pic"></button>
```

I also wanted to mention that for me an alternative text is not a description. Both of them are often mixed on social networks, maybe we should pay attention to this too when implementing those new features. 

If someone wants to describe their avatar or profile pic they can currently do it in their bio most of the time. Does it make sense to have a separate field from the bio, which would be dedicated to the description of the avatar? In this case, this text should be available in a paragraph close to the image. We could link both of them with the `aria-describedby` attribute. 

Inside the modal for example, the image is marked as decorative: 
```
<img role="presentation" alt="" title="" lang="" src="[...]" width="400" height="400" style="transform: scale(1) translate(0px); transform-origin: 0px 0px 0px;" draggable="false">
```

If we want to have this description, we could probably do something like this instead:
```
<img alt="[display name]" aria-describedby="avatar-desc">
<p id="avatar-desc">A white middle-aged woman with brown hair and eyes,... blah</p>
```
where [display name] should be replaced by the display name of the account.

For the banner, as far as I can tell, it is only displayed on the profile page and is never interactive. Most of the banners I have seen are decorative. But one could decide to convey information through this image. Here an alt attribute could be interesting.

**Comment by AlainVagner** [comment]
I have done some checks on other pages:
- on the home page with the main timeline, all avatars are in a link together with some related text: the name and handle. The image has an empty alt. The link has an non empty title attribute.  The empty alt attribute is ok, since the accessible name of the link is the name + handle. But the title attribute does not include the visible text. In the following example, the title attribute should contain `"Alain Vagner @AlainVagner"`
```
<a [...] href="/@AlainVagner" title="AlainVagner">
    <div class="account__avatar-wrapper"><div class="account__avatar" style="width: 46px; height: 46px;">
        <img src="[...].jpg" alt="">
    </div></div>
    <div class="account__contents"><span class="display-name">
        <bdi><strong class="display-name__html">Alain Vagner</strong></bdi> 
        <span class="display-name__account">@AlainVagner</span>
    </span></div>
</a>
``` 
- One of the issues I have seen is not related to the alternative of the image. In a timeline, on all avatars, there is a popup displayed on hover with some extra info and an unfollow button. This popup cannot be displayed when navigating with a keyboard.
- When a toot has been retooted, an avatar is displayed as an overlay on top of the main avatar of the toot. Both avatars have an non empty alt attribute. I would keep both alt attributes empty to avoid too much noise here. In the following example, the accessible name of the link is `"infobeautiful@vis.social AlainVagner Information Is Beautiful @infobeautiful@vis.social"` which seems redundant:
```
<a href="/@infobeautiful@vis.social" [...] title="infobeautiful@vis.social">
    <div class="status__avatar"><div class="account__avatar-overlay" style="width: 46px; height: 46px;"><div class="account__avatar-overlay-base"><div class="account__avatar" style="width: 36px; height: 36px;">
        <img src="[...].png" alt="infobeautiful@vis.social">
    </div></div>
    <div class="account__avatar-overlay-overlay"><div class="account__avatar" style="width: 24px; height: 24px;">
        <img src="[...].jpg" alt="AlainVagner">
    </div></div></div></div>
    <span class="display-name"><bdi><strong class="display-name__html">Information Is Beautiful</strong></bdi> <span class="display-name__account">@infobeautiful@vis.social</span></span>
</a>
```

**Comment by mayakarabula** [comment]
I have done some work already, but we can discuss on where and how to add this. Currently I have added it only to the profile page avatar and header, not on timelines etc as it was flagged to be an issue for screen readers.

**Comment by astranchet** [comment]
Nice start @mayakarabula 😊. In your draft, the alt text is displayed in the [components/avatar.tsx](https://github.com/mastodon/mastodon/pull/32593/files#diff-d5bce376f7848e7333d57ad5fe024354a3cafda4920274eadef754c917c101a8) template file. So, avatar alt text will be displayed everywhere this template is used (why is quite often, or maybe I'm missing something?). Maybe it should depend of the context, as @AlainVagner mentionned.

**Comment by mayakarabula** [comment]
@astranchet 
Ah of course, good catch. I wanted to add a boolean flag there whether to show the alt tag or not, now it would indeed show everywhere.

**Comment by pauloxnet** [comment]
Here to add that IMHO alt text would be useful at least in the profile page both for a avatar and header images.

**Comment by rkingett** [comment]
Having alt text in profile pictures in timelines will impede screen readers so keep it on the profile pages!

**Comment by MMachado05** [comment]
I don't mean to overcomplicate this issue, but does alt text change with locale? I post bilingually on Mastodon and frequently interact with some of the folks on the Costa Rican instance, and when I make a post primarily in Spanish, the alt text of the post is in Spanish as well. Could there be a way to mark different alt texts for different languages for those in similar circumstances?

As context, here's a Spanish post with Spanish alt text:

<img width="593" height="631" alt="A Mastodon post with text in Spanish, then in English. The attached images have alt text in Spanish." src="https://github.com/user-attachments/assets/aa87faa1-298e-40f5-81df-776fa7a21840" />

Whereas this is an English post with English alt text:

<img width="607" height="702" alt="A Mastodon post with text in English. The attached images have alt text in English." src="https://github.com/user-attachments/assets/f298958e-2c1d-426c-b6ef-f389b031ea13" />

**Comment by nclm** [comment]
This is now fixed with the new profile?

**Comment by jessebot** [comment]
I see this is available in the nightly docker image I'm using, however I'd like to suggest that the alt text for the banner and the profile picture be at least 250 characters instead of 150, so that people can fully describe themselves and any visible flags they have. With 150, I was not able to fully describe my profile picture, which as several pride flags in it in addition to my face. As an example, I'd want to say something like:

> Jesse, a pale olive skin tone Puerto Rican person with blue hair wearing an FFP3 mask surrounded by various pride flags: trans flag, progress flag, neurodivergent flag, and Puerto Rican heart flag.

This is because I have a few flags in the background of my profile picture. The result is that I have a total of 198 characters for my avatar and cannot fully describe it.

One more request while I'm at it, could this please work on hover? It'd be nice to view other people's alt text for their banner or avatar by hovering over it the same we can do with an image, because then we can also translate it if necessary.

**Comment by diondiondion** [comment]
@jessebot We'll keep the 150 characters limit for now but might revisit this in the future; feel free to open a new issue for this if you want.

**Comment by diondiondion** [comment]
I'll close this issue as starting with Mastodon 4.6, you'll be able to add/edit ALT text for your avatar and profile banner images (see our [blog post about the profile redesign](https://blog.joinmastodon.org/2026/03/a-redesign-for-profiles/#more-control-during-image-upload) for more details).

Alt attributes will only be populated to images on the profile page, not in feeds or other places where they could get in the way of browsing.
